cpp/src/arrow/dataset/scanner.cc

```diff
@@ -18,16 +18,28 @@
 #include "arrow/dataset/scanner.h"
 
 #include <algorithm>
+#include <memory>
 
 #include "arrow/dataset/dataset.h"
 #include "arrow/dataset/dataset_internal.h"
+#include "arrow/dataset/filter.h"
 #include "arrow/table.h"
 #include "arrow/util/iterator.h"
 
 namespace arrow {
 namespace dataset {
 
-RecordBatchIterator SimpleScanTask::Scan() { return MakeVectorIterator(record_batches_); }
+ScanOptions::ScanOptions()
+    : filter(scalar(true)), evaluator(ExpressionEvaluator::Null()) {}
```

Comment:
Keep it nullptr so we can distinguish between the user providing true or nothing.

Reply:
When would the user provide true? Why would we need to distinguish that? Again IMHO it's better to avoid nullables whenever possible.